#'Analyze the Data
#'
#'@description The goal is to output the information necessary for meta-analysis
#'@param phenofinal A data.frame of phenotype data that is subset to the samples
#'  among the matrix of processed beta-values. Be sure that the dataframe
#'  phenofinal includes the column 'Sex', which can only take three possible
#'  character values: 'Female', 'Male', or 'Unknown'. Missing values are not
#'  allowed. The data.frame must also have the column 'Basename'. If
#'  RestrictToSubset is TRUE, this data.frame must also include the column
#'  specified in the argument RestrictionVar
#'@param betafinal Matrix of the processed beta-values
#'@param array A character string of array type; options: "EPIC" or "450K"
#'@param Omega A required matrix of cell composition estimates generated by the
#'  preprocessingofData function
#'@param RestrictToSubset Logic statement (TRUE or FALSE), whether to restrict
#'  the analysis to a specific subset of the samples. The default is FALSE, in
#'  other words, do not restrict the analysis to a specific subset of the
#'  samples. If specified TRUE, must provide a character string to the arguments
#'  RestrictionVar and RestrictToIndicator; see those arguments for additional
#'  details.
#'@param RestrictionVar A character string indicating the column name for the
#'  variable in the phenofinal data.frame that will be used to subset the data.
#'  Must be specified if RestrictToSubset is TRUE; default is NULL
#'@param RestrictToIndicator A character string indicating the category of the
#'  RestrictionVar you would like to restrict the analysis to. This character
#'  string must equal one of the RestrictionVar categories. This argument is
#'  only required if RestrictToSubset = TRUE; default is NULL
#'@param Table1vars An character vector of variable column names to include in
#'  the descriptive Table 1. The column names must correspond the column names
#'  in the pData DataFrame of the GenomicRatioSet (e.g. c("Age", "BMI",
#'  "Smoke")); see the function loadingSamples for the new column names
#'@param StratifyTable1 Logic statement (TRUE or FALSE), whether to stratify the
#'  Table 1 data on a specific variable like infant sex or case/control status;
#'  default is FALSE
#'@param StratifyTable1var A character string indicating the column name for the
#'  variable on which to stratify the Table 1 data; default is FALSE. Must be
#'  specified if StratifyTable1 is TRUE
#'@param vartype A character string indicating how to model DNA methylation;
#'  four options: "ExposureCont", "ExposureCat", "OutcomeCont", "OutcomeBin". If
#'  "ExposureCont", the variable of interest (varofinterest) is modeled as a
#'  continuous exposure and assumed to be numeric. If "ExposureCat", the
#'  variable of interest is modeled as a categorical exposure and assumed to be
#'  a factor. If "OutcomeCont", the variable of interest is modeled as the
#'  outcome, and assumed to be continuous. If "OutcomeBin", the variable of
#'  interest is modeled as the outcome, and assumed to be binary (values are 0
#'  or 1). If DNAm is modeled as the exposure, the site-specific effect
#'  estimates will be for a 0.01 increase in beta-value
#'@param varofinterest A character string of the column name of the exposure or
#'  outcome of interest (e.g. "BWT"). The column name must correspond to a
#'  column name in the pData DataFrame of the GenomicRatioSet; see the function
#'  loadingSamples for the new column names
#'@param robust Logic statement (TRUE or FALSE), whether to run robust linear
#'  regressions. The default is TRUE
#'@param maxit If running robust linear regression models, the maximum number of
#'  iterations (default=100)
#'@param adjustmentvariables A character vector of variable column names to
#'  include in your multivariable adjusted models (e.g. c("Age","Gestage")). The
#'  column names must correspond the column names in the pData DataFrame of the
#'  GenomicRatioSet; see the function loadingSamples for the new column names
#'@param RunUnadjusted Run unadjusted models? Logic statement (TRUE or FALSE),
#'  default = TRUE
#'@param RunAdjusted Run adjusted models? Logic statement (TRUE or FALSE),
#'  default = TRUE
#'@param RunCellTypeAdjusted Run adjusted analysis, including cell composition?
#'  Logic statement (TRUE or FALSE), default = TRUE
#'@param RunSexSpecific Run models stratifying by infant sex? Logic statement
#'  (TRUE or FALSE), default = TRUE
#'@param RunCellTypeInteract Whether to run CellDMC to identify the specific
#'  cell type(s) driving the differential methylation. Logic statement (TRUE or
#'  FALSE), default = TRUE
#'@param runparallel Whether to run the site-specific analysis in parallel (TRUE
#'  or FALSE), default = FALSE
#'@param number_cores If running in parallel, a numeric string indicating the
#'  number of cores to run (e.g. 4), default = NULL
#'@param destinationfolder A character string indicating the location where
#'  files should be saved, e.g. "C:\\Home\\PACE\\BirthSize"
#'@param savelog logic; TRUE indicates to save a log of the functions run
#'  (default)
#'@param cohort A character string for the cohort's acronym (e.g. "HEBC")
#'@param analysisdate A character string indicating the date the analysis was
#'  run. Please specify in the form: YEARMONTHDAY, e.g. "20200205" for February
#'  5th 2020
#'@param analysisname An optional character string for the analysis name. This
#'  will be used to label the output directory and the RData files that contain
#'  the final results. This can be helpful to specify if running additional
#'  sensitivity analyses
#'@details  Data analysis steps include: \enumerate{ \item Create a Table 1
#'  descriptive table for the cohort \item Test the association between
#'  estimated cellular components and the exposure variable, the adjustment
#'  variables, and batch (if included)  \item Perform association analysis
#'  between variable of interest and each CpG site individually. If vartype is
#'  "ExposureCont", "ExposureCat", or "OutcomeCont", the association is
#'  estimated using linear regression modelling (if robust=TRUE, using the rlm()
#'  function in R; otherwise, using lm()). If vartype is "OutcomeBin", a
#'  logistic regression is used to estimate the association
#'  (glm(family="binomial")). Analyses performed include: \itemize{ \item
#'  Unadjusted analysis if RunUnadjusted=TRUE \item Analysis adjusting for
#'  specified adjustment variables if RunAdjusted=TRUE \item Analysis adjusting
#'  for specified adjustment variables and estimated cell composition if
#'  RunCellTypeAdjusted=TRUE \item If RunSexSpecific=TRUE, RunUnadjusted=TRUE,
#'  and there are at least 10 female infants in the dataset, unadjusted analysis
#'  among the female infants \item If RunSexSpecific=TRUE, RunAdjusted=TRUE, and
#'  there are at least 10 female infants in the dataset, analysis adjusting for
#'  specified adjustment variables among the female infants \item If
#'  RunSexSpecific=TRUE, RunCellTypeAdjusted=TRUE, and there are at least 10
#'  female infants in the dataset, analysis adjusting for specified adjustment
#'  variables and estimated cell composition among the female infants \item If
#'  RunSexSpecific=TRUE, RunUnadjusted=TRUE, and there are at least 10 male
#'  infants in the dataset, unadjusted analysis among the male infants \item If
#'  RunSexSpecific=TRUE, RunAdjusted=TRUE, and there are at least 10 male
#'  infants in the dataset, analysis adjusting for specified adjustment
#'  variables among the male infants \item If RunSexSpecific=TRUE,
#'  RunCellTypeAdjusted=TRUE, and there are more than 10 male infants in the
#'  dataset, analysis adjusting for specified adjustment variables and estimated
#'  cell composition among the male infants \item If RunCellTypeInteract=TRUE,
#'  output of CellDMC (from package EpiDISH) based on the adjusted model}  }
#'@return Outputs a list of data frames with the site-specific association
#'  results. Also creates a new folder in the Output folder for the specified
#'  exposure/outcome variable. The folder includes: \enumerate{\item Descriptive
#'  statistics of cohort (Table 1; csv file) \item Figures (PDFs/PNGs) of
#'  associations, including: \itemize{\item QQplot for each analysis \item
#'  Volcano plot for each analysis \item Scatterplot for the 20 most significant
#'  associations for each analysis} \item An RData file that includes a list of
#'  data frames with the site-specific association results \item An RData file
#'  that includes a list of the lambdas for each analysis}
#'@examples
#'\dontrun{
#'allresults<-dataAnalysis(phenofinal=pData(processedOut$mset),
#'                         betafinal=processedOut$processedBetas,
#'                         array="EPIC",
#'                         Omega=processedOut$Omega,
#'                         StratifyTable1=TRUE,
#'                         StratifyTable1var="Sex",
#'                         maxit=100,
#'                         Table1vars=c("Age","BMI","Smoke","Parity","Sex","Gestage","BWT"),
#'                         vartype="ExposureCont",
#'                         varofinterest="BWT",
#'                         adjustmentvariables=c("Age","Gestage"),
#'                         destinationfolder="H:/UCLA/PACE/Birthweight-placenta",
#'                         robust=TRUE,
#'                         RunUnadjusted=TRUE,
#'                         RunAdjusted=TRUE,
#'                         RunCellTypeAdjusted=TRUE,
#'                         RunSexSpecific=TRUE,
#'                         RunCellTypeInteract=TRUE,
#'                         RestrictToSubset=FALSE,
#'                         RestrictionVar=NULL,
#'                         RestrictToIndicator=NULL,
#'                         runparallel=FALSE,
#'                         number_cores=NULL,
#'                         savelog=TRUE,
#'                         cohort="HEBC",analysisdate="20200710",
#'                         analysisname="primaryBWT")
#'}

dataAnalysis<-function(phenofinal=NULL,betafinal=NULL,array="EPIC",
                       Table1vars=c("Age","BMI","Smoke","Parity","Sex","Gestage","BWT"),
                       StratifyTable1=FALSE,
                       StratifyTable1var=NULL,
                       maxit=100,
                       vartype="ExposureCont",
                       varofinterest=NULL,
                       adjustmentvariables=c("Age","Gestage"),
                       Omega=NULL,
                       RestrictToSubset=FALSE,
                       RestrictionVar=NULL,
                       RestrictToIndicator=NULL,
                       robust=TRUE,
                       RunUnadjusted=TRUE,
                       RunAdjusted=TRUE,
                       RunCellTypeAdjusted=TRUE,
                       RunSexSpecific=TRUE,
                       RunCellTypeInteract=TRUE,
                       runparallel=FALSE,
                       number_cores=NULL,
                       destinationfolder=NULL,savelog=TRUE,
                       cohort=NULL,analysisdate=NULL,
                       analysisname=NULL){


  ## Array must be "EPIC" or "450K"
  if(!(array %in% c("EPIC", "450K"))) stop("Array type must be 'EPIC' or '450K'")
  if(array=="450K") annotdat<-as.data.frame(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19))
  if(array=="EPIC") annotdat<-as.data.frame(getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19))

  if(RunCellTypeAdjusted & is.null(Omega)) stop("Cannot run analysis adjusting for cell composition if Omega is null")
  if(RunCellTypeInteract & is.null(Omega)) stop("Cannot evaluate interaction with cell composition if Omega is null")
  if(RunAdjusted & is.null(adjustmentvariables)) stop("If RunAdjusted=TRUE, must specify adjustment variables")
  if(runparallel & is.null(number_cores)) stop("If number_cores=TRUE, must specify the number of cores")

  if(!("Sex" %in% colnames(phenofinal))) stop("Variable 'Sex' is required and not in the dataframe phenofinal; see function arguments for specification")
  if(!("Basename" %in% colnames(phenofinal))) stop("Variable 'Basename' is required and not in the dataframe phenofinal")

  phenofinal$Sex<-as.character(phenofinal$Sex)
  initialF<-length(which(phenofinal$Sex=="Female"))
  initialM<-length(which(phenofinal$Sex=="Male"))
  initialUnknown<-length(which(phenofinal$Sex=="Unknown"))

  totalMandF<-initialF+initialM+initialUnknown
  if(totalMandF!=nrow(phenofinal)) stop("The number of males and females in the dataset does not match up
                                          with the number of samples. Be sure that the dataframe phenofinal
                                          includes the column 'Sex', which can only take three possible
                                          character values: 'Female', 'Male', or 'Unknown'. Missing values
                                          are not allowed")

  ## Making sure that annotdat has the same CpG loci as betafinal, and in the same order
  annotdat<-annotdat[rownames(betafinal),]

  ## Making sure that finalpheno has the same samples as betafinal, and in the same order
  finalpheno<-as.data.frame(phenofinal)
  rownames(finalpheno)<-as.character(finalpheno$Basename)

  ## checking overlapping samples for pData and beta matrix
  samplesinboth<-intersect(rownames(finalpheno),colnames(betafinal))
  if(length(samplesinboth)<10) stop("Less than 10 samples overlap between the pheno data.frame and the beta-value matrix")

  finalpheno<-finalpheno[samplesinboth,]
  betafinal<-betafinal[,samplesinboth]

  if(RestrictToSubset & is.null(RestrictionVar)) stop("If RestrictToSubset is TRUE, must specify variable used to subset the dataframe phenofinal; see function arguments for specification")
  if(RestrictToSubset & !is.null(RestrictionVar)) {
    if(!(RestrictionVar %in% colnames(finalpheno))) stop("If RestrictToSubset is TRUE, variable specified by the RestrictionVar argument is required in the dataframe phenofinal; see function arguments for specification")
  }

  ## If restricting to a specific subset
  if(RestrictToSubset){

    finalpheno$RestrictionVar<-as.character(finalpheno[,RestrictionVar])

    if(!(RestrictToIndicator %in% finalpheno$RestrictionVar)) stop("Specified indicator for subset group is not the column specified by RestrictToSubset")
    RestrictToIndicator<-as.character(RestrictToIndicator)

    numberinsubset<-length(finalpheno$RestrictionVar[which(finalpheno$RestrictionVar==RestrictToIndicator)])
    if(numberinsubset<10) stop("Less than 10 individuals of the specified subset")

    finalpheno<-finalpheno[which(finalpheno$RestrictionVar==RestrictToIndicator),]
    betafinal<-betafinal[,rownames(finalpheno)]

  }

  ## Making sure that Omega has the same samples as betafinal, and in the same order
  if(!is.null(Omega)) {
    if(is.matrix(Omega)){
      cat("Number of cell types:",ncol(Omega))
      Omega<-Omega[colnames(betafinal),]
    } else {
      cat("Number of cell types:",1)
      Omega<-Omega[colnames(betafinal)]
    }
  }

  ## Checking that varofinterest is the right class
  if(!(vartype %in% c("ExposureCont","ExposureCat","OutcomeCont","OutcomeBin"))) stop("vartype must be of type: 'ExposureCont','ExposureCat','OutcomeCont','OutcomeBin'")
  if(vartype=="ExposureCont" & !is.numeric(finalpheno[,varofinterest])) stop("If vartype is 'ExposureCont' varofinterest must be numeric")
  if(vartype=="ExposureCat" & !is.factor(finalpheno[,varofinterest])) stop("If vartype is 'ExposureCat' varofinterest must be a factor")
  if(vartype=="OutcomeCont" & !is.numeric(finalpheno[,varofinterest])) stop("If vartype is 'OutcomeCont' varofinterest must be numeric")
  if(vartype=="OutcomeBin" & !is.numeric(finalpheno[,varofinterest])) stop("If vartype is 'OutcomeBin' varofinterest must be numeric")
  if(vartype=="OutcomeBin"){
    tempoutcome<-finalpheno[,varofinterest]
    if(FALSE %in% (tempoutcome %in% c(0,1,NA))) stop("vartype is 'OutcomeBin' and varofinterest has value besides: 0, 1, or NA")
  }

  if(is.null(destinationfolder)) stop("Please specify a destination folder")

  ## Checking that base output folder is in the directory, if not there, creating it
  BaseOutputname<-paste(cohort,"_",analysisdate,"_Output",sep="")
  setwd(destinationfolder)
  tempdir<-dir()
  if(!(BaseOutputname %in% tempdir)) dir.create(file.path(destinationfolder, BaseOutputname), showWarnings = FALSE)

  ## Adding a subfolder for this variable of interest
  Outputname<-paste(BaseOutputname,"/",varofinterest,sep="")
  if(!is.null(analysisname)) Outputname<-paste(Outputname,"_",analysisname,sep="")

  dir.create(file.path(destinationfolder, Outputname), showWarnings = FALSE)

  if(RestrictToSubset){

    ## Adding an additional subfolder for specific subset
    Outputname<-paste(Outputname,"/",RestrictionVar,"_",RestrictToIndicator,sep="")
    dir.create(file.path(destinationfolder, Outputname), showWarnings = FALSE)

    ## Removing adjustment for subset variable because restricting to only one
    adjustmentvariables<-adjustmentvariables[which(adjustmentvariables!=RestrictionVar)]
    Table1vars<-Table1vars[which(Table1vars!=RestrictionVar)]

  }

  destinationfolder<-paste(destinationfolder,Outputname,sep="/")
  setwd(destinationfolder)

  if(savelog){

    sink()
    sink(type="message")

    con <- file("log_step4.log")
    sink(con, append=TRUE)
    sink(con, append=TRUE, type="message")

  }

  ## Checking dataset is ready
  if(FALSE %in% (adjustmentvariables %in% colnames(finalpheno))) stop("One of the adjustment variables is not in the dataset")
  if(!(varofinterest %in% colnames(finalpheno))) stop("Variable of interest is not in the dataset")
  if(FALSE %in% (Table1vars %in% colnames(finalpheno))) stop("One of the Table 1 variables is not in the dataset")

  cat("Variable of Interest:",varofinterest,"\n")
  cat("  Running robust models?",robust,"\n")
  cat("Creating Table 1...","\n")

  ##Adding outcome/exposure of interest and adjustment variables to Table 1 as well
  if(vartype=="OutcomeBin") {
    finalpheno$tempoutcomebin<-as.factor(finalpheno[,varofinterest])
    tempoutcomename<-paste(varofinterest,"asfactor",sep=".")
    colnames(finalpheno)[which(colnames(finalpheno)=="tempoutcomebin")]<-tempoutcomename
    Table1vars<-unique(c(Table1vars,adjustmentvariables,tempoutcomename))
  } else{
    Table1vars<-unique(c(Table1vars,adjustmentvariables,varofinterest))
  }

  ## make sure that the stratification variable is not among the Table1vars
  if(StratifyTable1) Table1vars<-Table1vars[-grep(StratifyTable1var,Table1vars)]

  subsetpheno <- finalpheno[,Table1vars]
  rownames(subsetpheno)<-as.character(finalpheno$Basename)

  ## complete case subset
  subsetphenocomplete <- finalpheno[,c("Basename",varofinterest,adjustmentvariables)]
  subsetphenocomplete <- na.omit(subsetphenocomplete)
  subsetphenocomplete <- as.character(subsetphenocomplete$Basename)

  ## Adding cell composition to Table 1 variables
  if(!is.null(Omega)) {
    if(is.matrix(Omega)){
      tempOmega<-as.data.frame(Omega)
      subsetpheno<-cbind(subsetpheno,tempOmega)
    } else {
      cat("Number of cell types:",1)
      subsetpheno<-cbind(subsetpheno,Omega=Omega)
    }
  }

  ## only run if subsetpheno is a dataframe
  if(is.data.frame(subsetpheno)){

    table.type = ifelse(lapply(subsetpheno, class) %in% c("numeric", "integer"),"Continuous","Categorical")
    table.type <- data.frame(Varname=colnames(subsetpheno),Type=table.type)
    subsetpheno$Basename <- finalpheno$Basename

    ## changing all variables to factors first before melting
    subsetpheno[sapply(subsetpheno, function(x) !is.factor(x))] <- lapply(subsetpheno[sapply(subsetpheno, function(x) !is.factor(x))], as.factor)

    if(StratifyTable1){

      if(!(StratifyTable1var %in% colnames(finalpheno))) stop("Missing stratification variable for Table 1")
      subsetpheno$StratifyTable1var<-finalpheno[,StratifyTable1var]
      subsetpheno$StratifyTable1var<-as.character(subsetpheno$StratifyTable1var)

      tempStratifyNumbers<-plyr::ddply(subsetpheno,.(StratifyTable1var),function(x) data.frame(Ncat=nrow(x)))
      tempStratifyNumbers.CompleteCase<-plyr::ddply(subsetpheno[rownames(subsetpheno) %in% subsetphenocomplete,],.(StratifyTable1var),function(x) data.frame(Ncat=nrow(x)))

      subsetpheno<-reshape::melt(subsetpheno,id=c("Basename","StratifyTable1var"))

    } else {
      subsetpheno<-reshape::melt(subsetpheno,id="Basename")
    }

    colnames(subsetpheno)[which(colnames(subsetpheno)=="variable")]<-"Varname"
    colnames(subsetpheno)[which(colnames(subsetpheno)=="value")]<-"CharValue"
    subsetpheno<-merge(subsetpheno,table.type,by="Varname")

    subsetpheno_cat<-subsetpheno[which(subsetpheno$Type=="Categorical"),]
    subsetpheno_cat$CharValue<-as.character(subsetpheno_cat$CharValue)
    subsetpheno_cat$Basename<-as.character(subsetpheno_cat$Basename)

    subsetpheno_cat_completecase<-subsetpheno_cat[subsetpheno_cat$Basename %in% subsetphenocomplete,]

    if(nrow(subsetpheno_cat)>0){
      if(StratifyTable1){

        Cat.Table.Summary<-plyr::ddply(subsetpheno_cat,.(Varname,StratifyTable1var),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=FALSE,title=tempname)
          tempout<-as.data.frame(tempout)
          colnames(tempout)<-c("Variable","Nmiss","Cat","Value")
          tempout
        })

        Cat.Table.Summary.CompleteCase<-plyr::ddply(subsetpheno_cat_completecase,.(Varname,StratifyTable1var),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=FALSE,title=tempname)
          tempout<-as.data.frame(tempout)
          colnames(tempout)<-c("Variable","Nmiss","Cat","Value")
          tempout
        })

      } else {

        Cat.Table.Summary<-plyr::ddply(subsetpheno_cat,.(Varname),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=FALSE,title=tempname)
          tempout<-as.data.frame(tempout)
          colnames(tempout)<-c("Variable","Nmiss","Cat","Value")
          tempout
        })

        Cat.Table.Summary.CompleteCase<-plyr::ddply(subsetpheno_cat_completecase,.(Varname),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=FALSE,title=tempname)
          tempout<-as.data.frame(tempout)
          colnames(tempout)<-c("Variable","Nmiss","Cat","Value")
          tempout
        })

      }

    } else {
      Cat.Table.Summary<-NULL
    }

    subsetpheno_cont<-subsetpheno[which(subsetpheno$Type=="Continuous"),]
    ## converting back from factor to numeric
    subsetpheno_cont$CharValue<-as.numeric(as.character(subsetpheno_cont$CharValue))
    subsetpheno_cont$Basename<-as.character(subsetpheno_cont$Basename)

    subsetpheno_cont_completecase<-subsetpheno_cont[subsetpheno_cont$Basename %in% subsetphenocomplete,]

    if(nrow(subsetpheno_cont)>0){
      if(StratifyTable1){

        Cont.Table.Summary<-plyr::ddply(subsetpheno_cont,.(Varname,StratifyTable1var),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=TRUE,title=tempname)
          tempout
        })
        colnames(Cont.Table.Summary)[3:6]<-c("Variable","Nmiss","Cat","Value")

        Cont.Table.Summary.CompleteCase<-plyr::ddply(subsetpheno_cont_completecase,.(Varname,StratifyTable1var),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=TRUE,title=tempname)
          tempout
        })
        colnames(Cont.Table.Summary.CompleteCase)[3:6]<-c("Variable","Nmiss","Cat","Value")


      } else {

        Cont.Table.Summary<-plyr::ddply(subsetpheno_cont,.(Varname),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=TRUE,title=tempname)
          tempout
        })
        colnames(Cont.Table.Summary)[2:5]<-c("Variable","Nmiss","Cat","Value")

        Cont.Table.Summary.CompleteCase<-plyr::ddply(subsetpheno_cont_completecase,.(Varname),function(x){
          tempname<-unique(as.character(x$Varname))
          tempout<-summarizeData(variable=x$CharValue,continuous=TRUE,title=tempname)
          tempout
        })
        colnames(Cont.Table.Summary.CompleteCase)[2:5]<-c("Variable","Nmiss","Cat","Value")

      }

    } else {
      Cont.Table.Summary<-NULL
    }

    allouttable<-rbind(Cont.Table.Summary,Cat.Table.Summary)
    rownames(allouttable)<-NULL
    allouttable$Varname<-NULL

    allouttable.CompleteCase<-rbind(Cont.Table.Summary.CompleteCase,Cat.Table.Summary.CompleteCase)
    rownames(allouttable.CompleteCase)<-NULL
    allouttable.CompleteCase$Varname<-NULL

    if(StratifyTable1){

      allouttable<-merge(allouttable,tempStratifyNumbers,by="StratifyTable1var")
      allouttable$StratifyTable1var<-paste(allouttable$StratifyTable1var," (n=", allouttable$Ncat,")",sep="")
      allouttable$Ncat<-NULL

      allouttable.CompleteCase<-merge(allouttable.CompleteCase,tempStratifyNumbers.CompleteCase,by="StratifyTable1var")
      allouttable.CompleteCase$StratifyTable1var<-paste(allouttable.CompleteCase$StratifyTable1var," (n=", allouttable.CompleteCase$Ncat,")",sep="")
      allouttable.CompleteCase$Ncat<-NULL

      allouttable$Cat<-as.character(allouttable$Cat)
      nmisstemp<-plyr::ddply(allouttable,.(Variable),function(x) data.frame(Nmiss=sum(as.numeric(x$Nmiss),na.rm=TRUE)))
      nmisstemp$Nmiss[which(nmisstemp$Nmiss==0)]<-""
      allouttable<-reshape::cast(allouttable,Variable+Cat~StratifyTable1var,value="Value")
      allouttable<-merge(nmisstemp,allouttable,by="Variable")

      allouttable.CompleteCase$Cat<-as.character(allouttable.CompleteCase$Cat)
      nmisstemp<-plyr::ddply(allouttable.CompleteCase,.(Variable),function(x) data.frame(Nmiss=sum(as.numeric(x$Nmiss),na.rm=TRUE)))
      nmisstemp$Nmiss[which(nmisstemp$Nmiss==0)]<-""
      allouttable.CompleteCase<-reshape::cast(allouttable.CompleteCase,Variable+Cat~StratifyTable1var,value="Value")
      allouttable.CompleteCase<-merge(nmisstemp,allouttable.CompleteCase,by="Variable")

    }

    allouttable$Variable[which(allouttable$Nmiss!="")]<-paste(allouttable$Variable[which(allouttable$Nmiss!="")]," (nmiss=",allouttable$Nmiss[which(allouttable$Nmiss!="")],")",sep="")
    allouttable$Nmiss<-NULL

    allouttable.CompleteCase$Variable[which(allouttable.CompleteCase$Nmiss!="")]<-paste(allouttable.CompleteCase$Variable[which(allouttable.CompleteCase$Nmiss!="")],
                                                                                        " (nmiss=",allouttable.CompleteCase$Nmiss[which(allouttable.CompleteCase$Nmiss!="")],")",sep="")
    allouttable.CompleteCase$Nmiss<-NULL

    OutputnameTable1<-paste(cohort,"_",analysisdate,"_",varofinterest,sep="")
    if(!is.null(analysisname)) OutputnameTable1<-paste(OutputnameTable1,"_",analysisname,sep="")

    write.csv(allouttable,paste(OutputnameTable1,"_Table_1.csv",sep=""),row.names=FALSE)

    write.csv(allouttable.CompleteCase,paste(OutputnameTable1,"_Table_1_Adjusted_Complete_Case.csv",sep=""),row.names=FALSE)

  }


  #################################################################################
  # See if cell-mixtures are associated with phenotypes/covariates of interest:
  
  if(!is.null(Omega)) {
    
    if(is.matrix(Omega)){
      
      if("Batch" %in% colnames(finalpheno)) {
        subsetpheno <- finalpheno[,unique(c(adjustmentvariables,varofinterest,"Batch"))]
      } else {
        subsetpheno <- finalpheno[,unique(c(adjustmentvariables,varofinterest))]
      }
      
      table.type = ifelse(lapply(subsetpheno, class) %in% c("numeric", "integer"),"Continuous","Categorical")
      table.type <- data.frame(Varname=colnames(subsetpheno),Type=table.type)
      
      subsetpheno$Basename <- finalpheno$Basename
      ## changing all variables to factors first before melting
      subsetpheno[sapply(subsetpheno, function(x) !is.factor(x))] <- lapply(subsetpheno[sapply(subsetpheno, function(x) !is.factor(x))], as.factor)
      
      subsetpheno<-reshape::melt(subsetpheno,id="Basename")
      colnames(subsetpheno)[which(colnames(subsetpheno)=="variable")]<-"Varname"
      colnames(subsetpheno)[which(colnames(subsetpheno)=="value")]<-"CharValue"
      subsetpheno<-merge(subsetpheno,table.type,by="Varname")
      OmegaDatFrame<-as.data.frame(Omega)
      OmegaDatFrame$Basename<-rownames(Omega)
      OmegaDatFrame<-reshape::melt(OmegaDatFrame,id="Basename")
      subsetpheno<-merge(subsetpheno,OmegaDatFrame,by="Basename")
      
      subsetpheno_cat<-subsetpheno[which(subsetpheno$Type=="Categorical"),]
      if(nrow(subsetpheno_cat)>0){
        
        error_message_cat <- function() data.frame(cat_name=NA,coef=NA,se=NA,pval=NA,warnings="This model failed")
        
        subsetpheno_cat$CharValue<-as.factor(subsetpheno_cat$CharValue)
        Cat.Table.p<-plyr::ddply(subsetpheno_cat,.(Varname,variable),function(x) {

          tryCatch(
            
            {
              
              tempframe<-data.frame(CellCompEst=x$value,CatVar=as.factor(as.character(x$CharValue)))
              mod <- rlm(CellCompEst~CatVar,data=tempframe)
              convergemessage<-ifelse(mod$converged,"none","'rlm' failed to converge")
              
              modout <- lmtest::coeftest(mod, vcov=vcovHC(mod, type="HC0"))
              modout <- modout[,]
              modout<-as.data.frame(modout)
              colnames(modout)<-c("coef","se","teststat","pval")
              modout$warnings<-convergemessage
              modout<-modout[2:nrow(modout),]
              modout<-cbind(cat_name=rownames(modout),modout)
              modout
              
            },
            
            error = function(e) error_message_cat()
            
          )
          
        })
        
        p<-ggplot(subsetpheno_cat,aes(CharValue,value))+
          geom_jitter(alpha=0.6)+
          geom_boxplot(outlier.shape=NA,fill=NA)+
          facet_grid(variable~Varname,scales="free")+
          theme_bw()+ylab("")+xlab("")+
          theme(axis.title=element_text(face="bold",size=14))
        
        ggsave(filename=paste(cohort,"_",analysisdate,"_CatCovariate_CellMix_Associations.png",sep=""),
               plot=p,width=10,height=10,units="in")
        write.csv(Cat.Table.p, paste(cohort,"_",analysisdate,"_CatCovariate_CellMix_Associations.csv",sep=""), na="NA")
        
      }
      
      subsetpheno_cont<-subsetpheno[which(subsetpheno$Type=="Continuous"),]
      if(nrow(subsetpheno_cont)>0){
        
        error_message_cont <- function() data.frame(coef=NA,se=NA,pval=NA,warnings="This model failed")
        
        ## changing all variables to factors first before melting
        subsetpheno_cont$CharValue<-as.numeric(as.character(subsetpheno_cont$CharValue))
        Cont.Table.p<-plyr::ddply(subsetpheno_cont,.(Varname,variable),function(x){
          
          tryCatch(
            
            {
              
              tempframe<-data.frame(CellCompEst=x$value,CharVal=x$CharValue)
              
              mod <- rlm(CellCompEst~CharVal,data=tempframe,maxit=maxit)
              convergemessage<-ifelse(mod$converged,"none","'rlm' failed to converge")
              
              modout <- lmtest::coeftest(mod, vcov=vcovHC(mod, type="HC0"))
              modout <- modout[,]
              modout<-as.data.frame(modout)
              colnames(modout)<-c("coef","se","teststat","pval")
              modout$warnings<-convergemessage
              modout[2,]
              
            },
            
            error = function(e) error_message_cont()
            
          )
          
        })
        
        p<-ggplot(subsetpheno_cont,aes(CharValue,value))+
          geom_point()+
          stat_smooth(method="lm",se=FALSE)+
          facet_grid(variable~Varname,scales="free")+
          theme_bw()+ylab("")+xlab("")+
          theme(axis.title=element_text(face="bold",size=14))
        
        ggsave(filename=paste(cohort,"_",analysisdate,"_ContCovariate_CellMix_Associations.png",sep=""),
               plot=p,width=10,height=10,units="in")
        write.csv(Cont.Table.p, paste(cohort,"_",analysisdate,"_ContCovariate_CellMix_Associations.csv",sep=""), na="NA")
        
      }
      
    }
    
  }
  
  #################################################################################
  # R-code for running models

  alldataout<-list()
  alllambda<-list()

  ## Run unadjusted model?
  if(RunUnadjusted){

    cat("Running Unadjusted Models...","\n")
    ## Unadjusted Models
    tempout<-runmodelfunction(methyldat=betafinal,vartype=vartype,robust=robust,
                              varofinterest=varofinterest,adjustmentvariables=NULL,
                              celladjust=FALSE,Omega=NULL,phenoframe=finalpheno,maxit=maxit,
                              runparallel=runparallel,number_cores=number_cores)
    
    tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Unadjusted_Analysis",
                                              vartype=vartype,finalpheno=finalpheno,
                                              annotdat=annotdat,betafinal=betafinal,
                                              varofinterest=varofinterest,
                                              cohort=cohort,analysisdate=analysisdate)
    
    alldataout$Unadjusted<-tempout_organizing$ModelResults
    alllambda$Unadjusted<-tempout_organizing$lambda

  }

  ## Run adjusted model?
  if(RunAdjusted){

    ## Adjusted model
    cat("Running Adjusted Models...","\n")
    cat("  Adjustment variables:",adjustmentvariables,"\n")
    tempout<-runmodelfunction(methyldat=betafinal,vartype=vartype,robust=robust,varofinterest=varofinterest,
                              adjustmentvariables=adjustmentvariables,
                              celladjust=FALSE,Omega=NULL,phenoframe=finalpheno,maxit=maxit,
                              runparallel=runparallel,number_cores=number_cores)
    
    tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Adjusted_Analysis",
                                              vartype=vartype,finalpheno=finalpheno,
                                              annotdat=annotdat,betafinal=betafinal,
                                              varofinterest=varofinterest,
                                              cohort=cohort,analysisdate=analysisdate)
    
    alldataout$Adjusted<-tempout_organizing$ModelResults
    alllambda$Adjusted<-tempout_organizing$lambda

  }

  ## Run model adjusting for cell types?
  if(RunCellTypeAdjusted){

    ## Adjusted model with Cell Type
    cat("Running Adjusted Models with Cell Type...","\n")
    if(class(Omega)[1]=="matrix"){
      cat("  Number of cell types:",ncol(Omega),"\n")
    } else {
      cat("  Number of cell types:",1,"\n")
    }

    tempout<-runmodelfunction(methyldat=betafinal,vartype=vartype,robust=robust,varofinterest=varofinterest,
                              adjustmentvariables=adjustmentvariables,
                              celladjust=TRUE,Omega=Omega,phenoframe=finalpheno,maxit=maxit,
                              runparallel=runparallel,number_cores=number_cores)
    
    tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Adjusted_Analysis_with_Cell_Type",
                                              vartype=vartype,finalpheno=finalpheno,
                                              annotdat=annotdat,betafinal=betafinal,
                                              varofinterest=varofinterest,
                                              cohort=cohort,analysisdate=analysisdate)
    
    alldataout$AdjustedwithCellType<-tempout_organizing$ModelResults
    alllambda$AdjustedwithCellType<-tempout_organizing$lambda

  }

  ## Run sex-specific models?
  if(RunSexSpecific){

    finalpheno$Sex<-as.character(finalpheno$Sex)
    numberF<-length(which(finalpheno$Sex=="Female"))
    numberM<-length(which(finalpheno$Sex=="Male"))

    ## If sex is an adjustment variable, need to drop it from
    ## adjustment variables for sex-stratified analyses
    if(!is.null(adjustmentvariables)){
      if("Sex" %in% adjustmentvariables){
        sexstratifiedadjustmentvars<-adjustmentvariables[-grep("Sex",adjustmentvariables)]
      } else {
        sexstratifiedadjustmentvars<-adjustmentvariables
      }
    } else {
      sexstratifiedadjustmentvars<-NULL
    }

    if(numberF>=10){

      tempsexbetas<-betafinal[,which(finalpheno$Sex=="Female")]
      tempsexpheno<-finalpheno[which(finalpheno$Sex=="Female"),]

      if(class(Omega)[1]=="matrix"){
        tempsexomega<-Omega[which(finalpheno$Sex=="Female"),]
      } else {
        tempsexomega<-Omega[which(finalpheno$Sex=="Female")]
      }

      if(RunUnadjusted){

        cat("Running Unadjusted Models for Females...","\n")
        ## Unadjusted model

        tempout<-runmodelfunction(methyldat=tempsexbetas,vartype=vartype,robust=robust,varofinterest=varofinterest,
                                  adjustmentvariables=NULL,
                                  celladjust=FALSE,Omega=NULL,phenoframe=tempsexpheno,maxit=maxit,
                                  runparallel=runparallel,number_cores=number_cores)
        
        tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Unadjusted_Analysis_Females",
                                                  vartype=vartype,finalpheno=tempsexpheno,
                                                  annotdat=annotdat,betafinal=tempsexbetas,
                                                  varofinterest=varofinterest,
                                                  cohort=cohort,analysisdate=analysisdate)
        
        alldataout$UnadjustedFemales<-tempout_organizing$ModelResults
        alllambda$UnadjustedFemales<-tempout_organizing$lambda
        
      }

      if(RunAdjusted){

        cat("Running Adjusted Models for Females...","\n")
        ## Adjusted model

        tempout<-runmodelfunction(methyldat=tempsexbetas,vartype=vartype,robust=robust,varofinterest=varofinterest,
                                  adjustmentvariables=sexstratifiedadjustmentvars,
                                  celladjust=FALSE,Omega=NULL,phenoframe=tempsexpheno,maxit=maxit,
                                  runparallel=runparallel,number_cores=number_cores)
        
        tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Adjusted_Analysis_Females",
                                                  vartype=vartype,finalpheno=tempsexpheno,
                                                  annotdat=annotdat,betafinal=tempsexbetas,
                                                  varofinterest=varofinterest,
                                                  cohort=cohort,analysisdate=analysisdate)
        
        alldataout$AdjustedFemales<-tempout_organizing$ModelResults
        alllambda$AdjustedFemales<-tempout_organizing$lambda

      }

      if(RunCellTypeAdjusted){

        cat("Running Adjusted Models with Cell Type for Females...","\n")
        ## Adjusted model with Cell Type
        tempout<-runmodelfunction(methyldat=tempsexbetas,vartype=vartype,robust=robust,varofinterest=varofinterest,
                                  adjustmentvariables=sexstratifiedadjustmentvars,
                                  celladjust=TRUE,Omega=tempsexomega,phenoframe=tempsexpheno,maxit=maxit,
                                  runparallel=runparallel,number_cores=number_cores)

        tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Adjusted_Analysis_with_Cell_Type_Females",
                                                  vartype=vartype,finalpheno=tempsexpheno,
                                                  annotdat=annotdat,betafinal=tempsexbetas,
                                                  varofinterest=varofinterest,
                                                  cohort=cohort,analysisdate=analysisdate)
        
        alldataout$AdjustedwithCellTypeFemales<-tempout_organizing$ModelResults
        alllambda$AdjustedwithCellTypeFemales<-tempout_organizing$lambda

      }
      
    }

    if(numberM>=10){

      tempsexbetas<-betafinal[,which(finalpheno$Sex=="Male")]
      tempsexpheno<-finalpheno[which(finalpheno$Sex=="Male"),]

      if(class(Omega)[1]=="matrix"){
        tempsexomega<-Omega[which(finalpheno$Sex=="Male"),]
      } else {
        tempsexomega<-Omega[which(finalpheno$Sex=="Male")]
      }

      if(RunUnadjusted){

        cat("Running Unadjusted Models for Males...","\n")
        ## Unadjusted model
        tempout<-runmodelfunction(methyldat=tempsexbetas,vartype=vartype,robust=robust,varofinterest=varofinterest,
                                  adjustmentvariables=NULL,
                                  celladjust=FALSE,Omega=NULL,phenoframe=tempsexpheno,maxit=maxit,
                                  runparallel=runparallel,number_cores=number_cores)
        
        tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Unadjusted_Analysis_Males",
                                                  vartype=vartype,finalpheno=tempsexpheno,
                                                  annotdat=annotdat,betafinal=tempsexbetas,
                                                  varofinterest=varofinterest,
                                                  cohort=cohort,analysisdate=analysisdate)
        
        alldataout$UnadjustedMales<-tempout_organizing$ModelResults
        alllambda$UnadjustedMales<-tempout_organizing$lambda

      }

      if(RunAdjusted){

        cat("Running Adjusted Models for Males...","\n")
        ## Adjusted model
        tempout<-runmodelfunction(methyldat=tempsexbetas,vartype=vartype,robust=robust,varofinterest=varofinterest,
                                  adjustmentvariables=sexstratifiedadjustmentvars,
                                  celladjust=FALSE,Omega=NULL,phenoframe=tempsexpheno,maxit=maxit,
                                  runparallel=runparallel,number_cores=number_cores)
        
        tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Adjusted_Analysis_Males",
                                                  vartype=vartype,finalpheno=tempsexpheno,
                                                  annotdat=annotdat,betafinal=tempsexbetas,
                                                  varofinterest=varofinterest,
                                                  cohort=cohort,analysisdate=analysisdate)
        
        alldataout$AdjustedMales<-tempout_organizing$ModelResults
        alllambda$AdjustedMales<-tempout_organizing$lambda

      }

      if(RunCellTypeAdjusted){

        cat("Running Adjusted Models with Cell Type for Males...","\n")
        ## Adjusted model with Cell Type
        tempout<-runmodelfunction(methyldat=tempsexbetas,vartype=vartype,robust=robust,varofinterest=varofinterest,
                                  adjustmentvariables=sexstratifiedadjustmentvars,
                                  celladjust=TRUE,Omega=tempsexomega,phenoframe=tempsexpheno,maxit=maxit,
                                  runparallel=runparallel,number_cores=number_cores)
        
        tempout_organizing<-organizing_DMP_output(ModelResults=tempout,ModelName="Adjusted_Analysis_with_Cell_Type_Males",
                                                  vartype=vartype,finalpheno=tempsexpheno,
                                                  annotdat=annotdat,betafinal=tempsexbetas,
                                                  varofinterest=varofinterest,
                                                  cohort=cohort,analysisdate=analysisdate)
        
        alldataout$AdjustedwithCellTypeMales<-tempout_organizing$ModelResults
        alllambda$AdjustedwithCellTypeMales<-tempout_organizing$lambda

      }

    }

  }

  ## Run cell-type interaction model?
  if(RunCellTypeInteract){

    ## Adjusted model
    cat("Running cell-type interaction models...","\n")

    ## This model cannot handle missing
    ## imputing missing with mean for that CpG locus
    tempbetasint<-betafinal
    kmiss <- which(is.na(tempbetasint), arr.ind=TRUE)
    tempbetasint[kmiss] <- rowMeans(tempbetasint, na.rm=TRUE)[kmiss[,1]]

    ## Must make sure there is no missing in the adjustment variables
    ## or the outcome variable
    tempdatanalysis<-finalpheno[,c("Basename",varofinterest,adjustmentvariables)]
    tempdatanalysis$OutputVar<-tempdatanalysis[,varofinterest]
    rownames(tempdatanalysis)<-as.character(tempdatanalysis$Basename)
    tempdatanalysis<-na.omit(tempdatanalysis)

    ## Make sure the beta matrix is the same samples
    tempbetasint<-tempbetasint[,rownames(tempdatanalysis)]

    ## Make sure the Omega matrix is the same samples
    tempOmegaint<-Omega[rownames(tempdatanalysis),]

    if(!is.null(adjustmentvariables)){

      tempmodinfo<-paste(adjustmentvariables,collapse = "+")
      tempmodinfo<-paste("~",tempmodinfo,sep="")
      tempmodinfo<-model.matrix(as.formula(tempmodinfo),data=tempdatanalysis)

      celldmc.o <- EpiDISH::CellDMC(beta.m=tempbetasint, pheno.v=tempdatanalysis$OutputVar, frac.m=tempOmegaint,
                           cov.mod=tempmodinfo,adjPThresh=0.05)

    } else {

      celldmc.o <- EpiDISH::CellDMC(beta.m=tempbetasint, pheno.v=tempdatanalysis$OutputVar, frac.m=tempOmegaint,
                           adjPThresh=0.05)

    }

    alldataout$CellInteraction<-celldmc.o

  }


  OutputnameFileName<-paste(cohort,"_",analysisdate,"_",varofinterest,sep="")
  if(!is.null(analysisname)) OutputnameFileName<-paste(OutputnameFileName,"_",analysisname,sep="")

  save(file=paste(OutputnameFileName,"_lambdas.RData",sep=""),compress=TRUE, list=c("alllambda"))
  save(file=paste(OutputnameFileName,"_allanalyses.RData",sep=""),compress=TRUE, list=c("alldataout"))
  if(savelog){
    sink()
    sink(type="message")
  }

  return(alldataout)
}


organizing_DMP_output<-function(ModelResults=NULL,ModelName=NULL,vartype=vartype,finalpheno=finalpheno,
                                annotdat=annotdat,betafinal=betafinal,varofinterest=varofinterest,
                                cohort=cohort,analysisdate=analysisdate){
  
  if(length(grep("pval",colnames(ModelResults)))>0){
    
    if(vartype=="ExposureCat"){
      
      pvalindexes<-grep("pval",colnames(ModelResults))
      catnames<-colnames(ModelResults)[pvalindexes]
      catnames<-sapply(strsplit(catnames,"_"), function(x) x[2])
      
      lambda<-as.list(rep(NA,length(catnames)))
      names(lambda)<-catnames
      
      for(i in catnames){
        
        ModelResultsv2<-ModelResults[,c("CpG","n",colnames(ModelResults)[grep(i,colnames(ModelResults),fixed = TRUE)])]
        colnames(ModelResultsv2)<-sapply(strsplit(colnames(ModelResultsv2),"_"), function(x) x[1])
        
        templambda = qchisq(median(ModelResultsv2$pval,na.rm=T), df = 1, lower.tail = F)/
          qchisq(0.5, 1)
        lambda[i]<-templambda
        ModelResultsplots<-na.omit(ModelResultsv2)
        
        if(nrow(ModelResultsplots)>20){
          runallplots(fitinfo=ModelResultsplots,ntop=20,annotdat=annotdat,betafinal=betafinal,
                      finalpheno=finalpheno,title=paste(ModelName,i,sep="_"),
                      varofinterest=varofinterest,vartype=vartype,
                      cohort=cohort,analysisdate=analysisdate)
        }
        
      }
      
    } else {
      
      lambda = qchisq(median(ModelResults$pval,na.rm=T), df = 1, lower.tail = F)/
        qchisq(0.5, 1)
      
      ModelResultsplots<-na.omit(ModelResults)
      if(nrow(ModelResultsplots)>20){
        runallplots(fitinfo=ModelResultsplots,ntop=20,annotdat=annotdat,betafinal=betafinal,
                    finalpheno=finalpheno,title=ModelName,
                    varofinterest=varofinterest,vartype=vartype,
                    cohort=cohort,analysisdate=analysisdate)
      }
      
    }
    
    outputlist<-list(ModelResults=ModelResults,lambda=lambda)
    
  } else {
    
    outputlist<-list(ModelResults=NA,lambda=NA)
    
  }
  
  return(outputlist)
  
}

error_message <- function() 'This model failed'

running_DMP_analysis<-function(x,inanalysis=NULL,Omega=NULL, modformula=NULL,vartype=NULL,
                               celladjust=NULL,robust=NULL,varofinterest=NULL,maxit=NULL){
  
  ## Scaling DNA methylation if it is the exposure
  if(vartype=="OutcomeCont" | vartype=="OutcomeBin"){
    x<-x*100
  }
  
  tempframe<-inanalysis
  tempframe$Methyl<-x
  
  if(celladjust) {
    if(class(Omega)[1]!="matrix") tempframe$tempOmega<-Omega
  }
  
  tempframe<-na.omit(tempframe)
  NinAnalysis<-nrow(tempframe)
  if(vartype=="ExposureCat") numbercategories<-length(levels(tempframe[,varofinterest]))
  
  if(celladjust) {
    
    if(class(Omega)[1]=="matrix") {
      
      tempOmega<-Omega[rownames(tempframe),]
      
    }
  }
  
  if(vartype=="OutcomeCont" | vartype=="ExposureCont" | vartype=="ExposureCat"){
    
    if(robust) {
      
      cf<-tryCatch(
        
        {
          
          mod <- rlm(as.formula(modformula),data=tempframe,maxit=maxit)
          convergemessage<-ifelse(mod$converged,"none","'rlm' failed to converge")
          
          modout <- lmtest::coeftest(mod, vcov=vcovHC(mod, type="HC0"))
          modout <- modout[,]
          modout<-as.data.frame(modout)
          modout$warnings<-convergemessage
          modout
          
        },
        
        error = function(e) error_message()
        
      )
      
    } else {
      
      
      cf<-tryCatch(
        
        {
          
          mod <- lm(as.formula(modformula),data=tempframe)
          modout <- summary(mod)$coef
          modout<-as.data.frame(modout)
          modout$warnings<-"none"
          modout
          
        },
        
        error = function(e) error_message()
        
      )
      
    }
    
    if(class(cf)=="data.frame"){
      
      if(vartype=="ExposureCat"){
        
        ## if number of categories is greater than two, output all exposure of interest coefficients
        if(numbercategories>2){
          
          outtempv1<-as.matrix(cf[c(2:numbercategories),c(1:4)])
          colnames(outtempv1)<-c("coef","se","teststat","pval")
          
          outtemp<-NULL
          for(p in 1:nrow(outtempv1)){
            
            tempinfo<-outtempv1[p,]
            names(tempinfo)<-paste(colnames(outtempv1),rownames(outtempv1)[p],sep="_")
            outtemp<-c(outtemp,tempinfo)
            
          }
          
          outtemp<-as.data.frame(matrix(outtemp,nrow=1,dimnames =list(NULL,names(outtemp))))
          outtemp<-cbind(n=NinAnalysis,outtemp,warnings=unique(cf$warnings))
          
        } else {
          
          outtemp<-as.data.frame(matrix(cf[2,],nrow=1,dimnames =list(NULL,colnames(cf))))
          colnames(outtemp)<-paste(c("coef","se","teststat","pval","warnings"),rownames(cf)[2],sep="_")
          outtemp<-cbind(n=NinAnalysis,outtemp)
          
        }
        
      } else {
        
        outtemp<-as.data.frame(matrix(cf[2,],nrow=1,dimnames =list(NULL,colnames(cf))))
        colnames(outtemp)<-c("coef","se","teststat","pval","warnings")
        outtemp<-cbind(n=NinAnalysis,outtemp)
        
      }
      
    } else {
      
      outtemp<-data.frame(n=NinAnalysis,coef=NA,se=NA,teststat=NA,pval=NA,warnings="This model failed")
      
    }
    
  }  
  
  if(vartype=="OutcomeBin"){
    
    Tempoutcome<-tempframe[,varofinterest]
    Ncases<-length(Tempoutcome[which(Tempoutcome==1)])
    Ncontrols<-length(Tempoutcome[which(Tempoutcome==0)])
    mod = glm(as.formula(modformula),data=tempframe,family = "binomial")
    convergemessage<-ifelse(mod$converged,"none","glm failed to converge")
    cf<-summary(mod)$coef
    outtemp<-as.data.frame(matrix(cf[2,],nrow=1,dimnames =list(NULL,colnames(cf))))
    colnames(outtemp)<-c("coef","se","zval","pval")
    outtemp<-cbind(n=NinAnalysis,n_cases=Ncases,n_controls=Ncontrols,outtemp,warnings=convergemessage)
    
  }
  
  return(outtemp) 
  
}

runmodelfunction<-function(methyldat=NULL,vartype=NULL,varofinterest=NULL,adjustmentvariables=NULL,
                           celladjust=NULL,Omega=NULL,phenoframe=NULL,robust=NULL,maxit=NULL,
                           runparallel=NULL,number_cores=NULL){

  ## Reducing phenotype file to essential variables
  inanalysis<-phenoframe[,c("Basename",varofinterest,adjustmentvariables)]

  if(vartype=="ExposureCont" | vartype=="ExposureCat"){ ## If DNA methylation is the outcome

    modformula<-paste("Methyl~",varofinterest,sep="")

  } else {  ## If DNA methylation is the exposure

    modformula<-paste(varofinterest,"~Methyl",sep="")

  }

  ## Adding adjustment variables to model (if not NULL)
  if(!is.null(adjustmentvariables)) modformula<-paste(modformula,"+",paste(adjustmentvariables,collapse = "+"),sep="")

  ## Adding cell composition estimate to the model if celladjust is TRUE
  if(celladjust) {

    modformula<-paste(modformula,"+ tempOmega",sep="")

    if(class(Omega)[1]=="matrix") {

      rownames(Omega)<-rownames(inanalysis)

    }

  }

  ## Site-specific analysis
  
  if(runparallel){
    
    cl <- parallel::makeCluster(number_cores)
    doParallel::registerDoParallel(cl)
    parallel::clusterExport(cl,c("inanalysis","modformula","maxit","rlm","Omega","error_message",
                       "vartype","celladjust","robust","varofinterest","running_DMP_analysis","vcovHC","coeftest"),envir=environment())
    
    DMP_list_out <- parallel::parApply(cl,methyldat,1, function(x) running_DMP_analysis(x=x,inanalysis=inanalysis,Omega=Omega,
                                                                                        modformula=modformula,vartype=vartype,
                                                                                        celladjust=celladjust,robust=robust,
                                                                                        varofinterest=varofinterest,maxit=maxit))
    stopCluster(cl)
    
  } else {
    
    DMP_list_out <- apply(methyldat,1, function(x) running_DMP_analysis(x,inanalysis=inanalysis,Omega=Omega,
                                                                        modformula=modformula,vartype=vartype,
                                                                        celladjust=celladjust,robust=robust,
                                                                        varofinterest=varofinterest,maxit=maxit))
    
  }
  
  tempnumbercols<-sapply(DMP_list_out,length)
  
  if(diff(range(tempnumbercols))>0){
    
    colnamesnew<-DMP_list_out[which(tempnumbercols==max(tempnumbercols))]
    colnamesnew<-colnames(colnamesnew[[1]])
    tempinputframe<-as.data.frame(matrix(NA,ncol=length(colnamesnew),nrow=1))
    colnames(tempinputframe)<-colnamesnew
    tempinputframe$warnings<-'This model failed'
    
    DMP_list_out<-lapply(DMP_list_out,function(x){
      templength<-ncol(x)
      if(templength==min(tempnumbercols)){
        tempinputframe
      } else {
        x
      }
    })
    
  }
  
  outresults<-data.frame(matrix(NA,ncol=length(DMP_list_out[[1]]),nrow=length(DMP_list_out)))
  colnames(outresults)<-colnames(DMP_list_out[[1]])
  rownames(outresults)<-names(DMP_list_out)
  for(i in 1:ncol(outresults)) outresults[,i]<-unlist(lapply(DMP_list_out,function(x)x[i]))
  outresults<-cbind(CpG=rownames(outresults),outresults)
  
  return(outresults)

}

qqnormplot<-function(fitinfo=NULL,title=NULL){
  o <- -log10(sort(fitinfo$pval,decreasing=F))
  e <- -log10( 1:length(o)/length(o))
  qqdata<-data.frame(o=o,e=e)
  ggplot(qqdata,aes(e,o))+
    geom_point()+
    geom_abline(intercept=0,slope=1,col="#d7301f")+
    xlab(expression(Expected~(-log[10](pval))))+ylab(expression(Observed~(-log[10](pval))))+
    theme_bw()+
    ggtitle(title)+
    theme(axis.title=element_text(face="bold",size=12),
          plot.title=element_text(face="bold",size=18,hjust=0))
}

volcanoplot<-function(fitinfo=NULL,title=NULL){

  fitinfo$fdr<-p.adjust(fitinfo$pval,method="fdr")
  if(min(fitinfo$fdr)<0.05){
    subsetsig<-fitinfo[which(fitinfo$fdr<0.05),]
    fdrcut<-subsetsig$pval[which.min(0.05-subsetsig$fdr)]
  } else {
    fdrcut<-0
  }

  ggplot(fitinfo,aes(coef,-log10(pval)))+
    geom_point(alpha=0.6)+
    geom_hline(yintercept=-log10(fdrcut),col="#78c679")+ #fdr, green
    geom_hline(yintercept=-log10(0.05/nrow(fitinfo)),col="#e7298a")+ #bonferroni, magenta
    xlab("Coefficient")+ylab(expression(-log[10](pval)))+
    theme_bw()+
    ggtitle(title)+
    theme(axis.title=element_text(face="bold",size=12),
          plot.title=element_text(face="bold",size=18,hjust=0))
}

mostsig<-function(fitinfo=NULL,ntop=NULL,annotdat=NULL,betafinal=NULL,
                  finalpheno=NULL,title=NULL,varofinterest=NULL,vartype=NULL){

  fitinfo<-fitinfo[order(fitinfo$pval),]
  topsig<-fitinfo[1:ntop,]
  topsigannot<-annotdat[as.character(topsig$CpG),]
  topsigannot$CpG<-as.character(topsigannot$Name)
  fitinfo<-merge(fitinfo,topsigannot,by="CpG")
  fitinfo$UCSC_RefGene_Name<-as.character(fitinfo$UCSC_RefGene_Name)
  fitinfo$UCSC_RefGene_Name<-sapply(strsplit(fitinfo$UCSC_RefGene_Name,";"),function(x)paste(unique(x),collapse=";"))
  fitinfo$label<-paste(fitinfo$CpG,"; p=",signif(fitinfo$pval, digits = 3),"\n",fitinfo$chr,":",fitinfo$pos,"; gene=",fitinfo$UCSC_RefGene_Name,sep="")

  fitbetas<-betafinal[as.character(topsig$CpG),]
  colnames(fitbetas)<-as.character(finalpheno$Basename)
  fitbetas<-reshape::melt(fitbetas)
  colnames(fitbetas)<-c("CpG","Basename","betavalue")
  finalpheno$ExposureVar<-finalpheno[,varofinterest]
  fitpData<-finalpheno[,c("Basename","ExposureVar","Sex")]
  fitpData<-merge(fitpData,fitbetas,by="Basename")
  finalplotting<-merge(fitinfo,fitpData,by="CpG")
  finalplotting<-finalplotting[order(finalplotting$pval),]
  labelorder<-unique(as.character(finalplotting$label))
  finalplotting$label<-as.factor(finalplotting$label)
  finalplotting$label<-factor(finalplotting$label,levels=labelorder)

  if(vartype=="ExposureCont" | vartype=="OutcomeCont"){
    ggplot(finalplotting,aes(ExposureVar,betavalue))+
      geom_point()+
      facet_wrap(~label,scales="free",ncol=4)+
      theme_bw()+
      stat_smooth(method="lm",se=FALSE)+
      ylab("Beta-value")+xlab(varofinterest)+
      ggtitle(title)+
      theme(axis.title=element_text(face="bold",size=12),
            plot.title=element_text(face="bold",size=18,hjust=0))
  } else {
    ggplot(finalplotting,aes(as.factor(ExposureVar),betavalue))+
      geom_boxplot(outlier.shape = NA)+
      geom_jitter(width = 0.2)+
      facet_wrap(~label,scales="free",ncol=4)+
      theme_bw()+
      ylab("Beta-value")+xlab(varofinterest)+
      ggtitle(title)+
      theme(axis.title=element_text(face="bold",size=12),
            plot.title=element_text(face="bold",size=18,hjust=0))
  }
}

runallplots<-function(fitinfo=NULL,ntop=NULL,annotdat=NULL,betafinal=NULL,
                      finalpheno=NULL,title=NULL,
                      varofinterest=NULL,vartype=NULL,
                      cohort=NULL,analysisdate=NULL){

  p<-qqnormplot(fitinfo=fitinfo,title=title)
  ggsave(filename=paste(cohort,"_",analysisdate,"_QQnorm_",title,".png",sep=""),
         plot=p,width=10,height=10,units="in")

  p<-volcanoplot(fitinfo=fitinfo,title=title)
  ggsave(filename=paste(cohort,"_",analysisdate,"_Volcano_",title,".png",sep=""),
         plot=p,width=10,height=10,units="in")

  p<-mostsig(fitinfo=fitinfo,ntop=20,annotdat=annotdat,betafinal=betafinal,
             finalpheno=finalpheno,title=title,varofinterest=varofinterest,vartype=vartype)
  ggsave(filename=paste(cohort,"_",analysisdate,"_Most_Sig_",title,".pdf",sep=""),plot=p,width=12,height=10,units="in")

}

summarizeData<-function(variable=NULL,continuous=TRUE,title=NULL){

  tempvariable<-variable

  if(continuous){

    tempvariable<-as.numeric(tempvariable)
    tempmedian<-format(round(median(tempvariable,na.rm=TRUE),2),nsmall=2)
    temprange<-trimws(format(round(range(tempvariable,na.rm = TRUE),2),nsmall=2))

    tempmean<-format(round(mean(tempvariable,na.rm=TRUE),2),nsmall=2)
    tempSD<-trimws(format(round(sd(tempvariable,na.rm = TRUE),2),nsmall=2))

    nmiss<-length(tempvariable[is.na(tempvariable)])

    ## note if any missing data and create output
    infoout<-c(title,nmiss,"",paste(tempmedian," (",temprange[1],"-",temprange[2],"); ",tempmean," (",tempSD,")",sep=""))

  } else {

    nmiss<-length(tempvariable[is.na(tempvariable)])
    tempvariable<-na.omit(tempvariable)
    tempvariable<-as.factor(tempvariable)
    tempcounts<-table(tempvariable)
    temppercent<-paste(round((table(tempvariable)/length(tempvariable))*100,2),"%",sep="")
    catout<-as.matrix(data.frame(cat=names(tempcounts),label=paste(tempcounts," (",temppercent,")",sep="")))
    #catout<-cbind("","",catout)
    #infoout<-rbind(c(title,"","",""),catout)
    infoout<-cbind(title,"",catout)
    infoout<-rbind(c(title,nmiss,"",""),infoout)

  }

  return(infoout)

}

